### ISSUE TITLE
Stuttery Audio playback

### ISSUE THREAD

**Issue by A-Cloud-Ninja** [bkacjios/lua-mumble#5]
I have the project working well, the chatbot features we need are fine, however we want a new music bot and i've gotten the functionality working, however the audio sometimes stutters, and it makes it a little awkward. Both while connecting to the server over a network, and while connecting internally.

**Comment by CaptainZidgel** [comment]
I've started experimenting with the sound capabilities of the library also and found the playback to be stuttery and pitched downward.
Here's a sound comparison: https://youtu.be/FSerBYyitLg
I don't know if it's the library or my network connection, but it's not the server bandwidth. (I found a public server with larger bandwidth than mine to test).

**Comment by bkacjios** [comment]
The audio files bitrate needs to be at 48000Khz for it to sound right since there is no re-sampling done.

As far as stuttering goes, maybe your server doesn't have the bandwidth setting high enough. I could probably port over my auto adjusting bitrate stuff from my Lumble project, so it will set the bitrate to comply with the servers settings on connect.

**Comment by A-Cloud-Ninja** [comment]
The playback is just stuttery for me, i have files properly converted to the bitrate and what not
ffmpeg command for that for anyone wondering: `"ffmpeg -i input_file.mp3 -ar 48000 -ac 1 -y "outputfile.ogg"`

The only issue is during playback there's weird artifacting in the audio

**Comment by A-Cloud-Ninja** [comment]
If we have to, you're more than welcome to not only investigate my code, but join our server and investigate as well.

Also for reference, our server is set to 256k, so it shouldn't be that our bitrate is too low, because it's running on a 40 gigabit VPS.

**Comment by bkacjios** [comment]
Weird. I've always had a bit of audio stutter as well, but it was never really that big of a deal for me. I think most of it comes down to it using the TCP pipe for audio rather than UDP; like normal clients. It could also be that the opus decoder is taking too long at times. When I had this running on a raspberry pi I noticed significant stuttering, but when I ran it on more powerful hardware the stuttering was less noticeable.

I'm not really sure how to make it stutter any less.

**Comment by A-Cloud-Ninja** [comment]
How is your other library, lumble, with this? Does it also have stuttering?

**Comment by bkacjios** [comment]
I feel like it's better, but there is still definite stutter.

**Comment by bkacjios** [comment]
I committed a pretty big change to the internal system of the module and I think it should help a lot with audio stuttering. This change will require a new dependency `libev-dev` and removes the functions `client:update()` and `client:sleep()` in favor of a blocking `mumble.loop()` call instead. [I updated the examples to reflect this.](https://github.com/bkacjios/lua-mumble/blob/4151728f6948d25004873e579b961baa6ac4b470/examples/afk.lua#L50)

Next I will be working on making the encoder auto adjust its bitrate based on the servers max_bandwidth config.
